### PR TITLE
Add UART buffer from old Zephyr code

### PR DESCRIPTION
**Pull request by jetremblay**

### CODE REVIEW

**Comment by jetremblay** [comment]
I modified the code slightly to work with newer mbed and removed the need to attach the IRQ handler in the user code.

File: utils/uart_buffer.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	bool txFull() const {
+		return txBuffer.full();
+	}
+
+	int write(uint8_t* msg, uint8_t len) {
+		int i;
+
+		// Message will be dropped if not enough space
+		// TODO: check actual remaining space, and compare with size of message
+		if(!txBuffer.full())
```
Comment (ducky64) [inline utils/uart_buffer.h:74]: I'd argue that writes should be blocking, but you could have a different API that checks for available space and returns whether the write was successful

File: utils/uart_buffer.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		if(txBuffer.empty())
+		{
+			// Do nothing
+			uart.attach(NULL,Serial::TxIrq);
+			idle = true;
+		}
+	}
+
+private:
+	calsol::util::CircularBuffer<uint8_t, TXSize> txBuffer;
```
Comment (ducky64) [inline utils/uart_buffer.h:108]: Not sure if circular buffer is concurrency-safe. In particular, the buffer elements are not volatile - is there a chance they could be reordered incorrectly wrt the start/end pointers?

File: utils/uart_buffer.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ *    // 32 message buffer
+ *    UARTTXBuffer<32> uartBuffer(uart);
+ *
+ *    void handleUartMessage() {
+ *        uartBuffer.handleIrq();
+ *    }
+ *
+ *    void setup() {
+ *       // Do your other setup
+ *       uartBuffer.attach(handleUartMessage);
```
Comment (ducky64) [inline utils/uart_buffer.h:28]: This is no longer necessary right? (and all related code that explicitly deals with IRQs)

File: utils/uart_buffer.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		return txBuffer.full();
+	}
+
+	int write(uint8_t* msg, uint8_t len) {
+		int i;
+
+		// Message will be dropped if not enough space
+		// TODO: check actual remaining space, and compare with size of message
+		if(!txBuffer.full())
+		{
```
Comment (ducky64) [inline utils/uart_buffer.h:75]: Style is inconsistent: there should be a space after the if, and the curly brace shouldn't get its own line

File: utils/uart_buffer.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+			for (i = 0; i < len; i++) {
+				txBuffer.write(msg[i]);
+			}
+		}
+
+		// Attach interrupt if first value, this will start transmission
+		if(idle) {
+			idle = false;
+			uart.attach(callback(this,&UARTTXBuffer<TXSize>::handleIrq),Serial::TxIrq);
```
Comment (ducky64) [inline utils/uart_buffer.h:84]: Does attach need to be called repeatedly and registered/deregistered? Can you call attach just once in the constructor, and only write the first character if not idle?

**Review by ducky64 — COMMENTED** [review]
Also check out BufferedSerial (third party mbed library), which I think does something similar (irq driven rx/tx)

**Review by ducky64 — COMMENTED** [review]
Also style-wise, why tabs instead of spaces?

**Comment by jetremblay** [comment]
There is also UARTSerial included in mbed, which appears to do the same thing. I will try that first.

**Comment by jetremblay** [comment]
UARTSerial duplicates this so I won't merge.

**Pull request closed without merging**